### PR TITLE
Relabel job_name as a meta label

### LINKED ISSUE #3033: How to get the original job tag?
**What did you do?**

I create the following ServiceMonitor

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    k8s-app: test
  name: test
  namespace: default
spec:
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 15s
    port: metrics
    relabelings:
    - sourceLabels:
      - job
      targetLabel: prometheus_job_name
  namespaceSelector: {}
  selector:
    matchLabels:
      app: test
```

and get the following prometheus config:

```yaml
scrape_configs:
- job_name: default/test/0
  honor_timestamps: true
  scrape_interval: 15s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - role: endpoints
    namespaces:
      names:
      - default
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_app]
    separator: ;
    regex: test
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: metrics
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Node;(.*)
    target_label: node
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Pod;(.*)
    target_label: pod
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: job
    replacement: ${1}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: endpoint
    replacement: metrics
    action: replace
  - source_labels: [job]
    separator: ;
    regex: (.*)
    target_label: prometheus_job_name
    replacement: $1
    action: replace
```

I need to get the original job_name as a tag `prometheus_job_name`, which should results to `default/test/0` here, but as the config show, relabelings defined in ServiceMonitor object are appended after built-in configs, so when the last relabeling executing, `job` has already been modified to `test`, I have no where to read `default/test/0`

My question: Is there any way the get the original `job` tag `default/test/0` ? 

or how can we change the order of relabelings? So I can read it before modified.

**Environment**

* Prometheus Operator version:

爱奇艺20200220225027

* Kubernetes version information:

1.15.0

### PR DESCRIPTION
**Pull request by lambdanis**
Fixes #3033

Keeping job_name as a meta label (__tmp_prometheus_job_name) allows users to keep it after further relabeling, e.g. for debugging purposes.

I feel this should be documented somewhere, but not sure where. Any suggestions?

**Release Note Template (will be copied)**
```release-note:enhancement
Keep job_name from generated scrape config as __tmp_prometheus_job_name label.
```

### CODE REVIEW

**Comment by s-urbaniak** [comment]
Thank you for the contribution! LGTM based on the comments posted in https://github.com/prometheus-operator/prometheus-operator/issues/3033#issuecomment-638722970 but I would like to have another pair of eyes, cc @prometheus-operator/kube-prometheus-reviewers

**Comment by lambdanis** [comment]
Thanks, any idea why e2e tests are failing? I tried to run them locally (minikube with k8s 1.16), but I'm getting errors like
```=== RUN   TestAlertmanagerInstanceNs/AllNs
    alertmanager_instance_namespaces_test.go:53: initialize Alertmanager CRD: alertmanagers CRD not ready: timed out waiting for Custom Resource: failed to list CRD: Get "https://192.168.49.2:8443/apis/monitoring.coreos.com/v1/alertmanagers": stream error: stream ID 601; INTERNAL_ERROR
--- FAIL: TestAlertmanagerInstanceNs (15.71s)
    --- FAIL: TestAlertmanagerInstanceNs/AllowList (5.50s)
    --- FAIL: TestAlertmanagerInstanceNs/DenyNs (5.49s)
    --- FAIL: TestAlertmanagerInstanceNs/AllNs (4.71s)```

File: pkg/prometheus/promcfg.go
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1076]: As per https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config, the `__meta` prefix is reserved for Prometheus service discovery. We need to use `__tmp` rather.

```suggestion
		{Key: "target_label", Value: "__tmp_prometheus_job_name"},
```

**Review by simonpasquier — COMMENTED** [review]
I've relaunched the e2e tests, lets see if it's a flake or not. As for documentation, I'm not sure where it should go either...

File: pkg/prometheus/promcfg.go
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:595]: Another option would be to initialize `relabelings` like this:

```
relabelings = []yaml.MapSlice{
  {Key: "source_labels", Value: []string{"job"}},
  {Key: "target_label", Value: "__tmp_prometheus_job_name"}
}
```

which would avoid the need for the `jobName` variable.

**Review by simonpasquier — COMMENTED** [review]
lgtm. I have a small suggestion but I'm ok merging this as-is.

File: pkg/prometheus/promcfg.go
Comment (QuentinBisson) [inline pkg/prometheus/promcfg.go:494]: Would it make sense to extract this to avoid duplication of code?

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:494]: if it makes sense yes but I wouldn't block the PR on it.

**Comment by simonpasquier** [comment]
You can add some documentation to the `relabelings` field so users can know that `__tmp_prometheus_job_name` exists.
https://github.com/prometheus-operator/prometheus-operator/blob/d10bc67f226e7c23efc1202928697d221478b7b9/pkg/apis/monitoring/v1/types.go#L782
https://github.com/prometheus-operator/prometheus-operator/blob/d10bc67f226e7c23efc1202928697d221478b7b9/pkg/apis/monitoring/v1/types.go#L851

**Comment by lambdanis** [comment]
> You can add some documentation to the `relabelings` field so users can know that `__tmp_prometheus_job_name` exists.

I added some docs there. What would you say to a separate file describing all labels automatically added by Prometheus Operator? Or maybe should that live in `design.md`? I didn't find such list anywhere and IMO it would be useful, I could attempt such docs in a separate PR.

**Review by simonpasquier — APPROVED** [review]
thanks! (and sorry for the delay)

**Pull request merged**
